### PR TITLE
Import a CDISC ODM to create an app

### PR DESCRIPTION
**Pull request by kaapstorm**
A work in progress. Needs some tests, and maybe a rebase to clean up the commits. 

This builds on PR #8953 and should provide some context to XFormBuilder.

@millerdev @snopoke cc @orangejenny @czue

### CODE REVIEW

File: custom/openclinica/management/commands/odm_to_app.py [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+        else:
+            module.add_insert_form(None, self._form)
+
+    def build_xform(self):
+        xform = XFormBuilder()
+        for ig in self.iter_item_groups():
+            data_type = 'repeatGroup' if self.is_repeating else 'group'
+            xform.add_question(ig.question_name, ig.question_label, data_type)
+            for item in ig.iter_items():
+                xform.add_question(item.question_name, item.question_label, DATA_TYPES[item.data_type], group=ig.oid)
```
Comment (dimagimon) [inline custom/openclinica/management/commands/odm_to_app.py]: E501 line too long (117 > 115 characters)

File: custom/openclinica/management/commands/odm_to_app.py [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+        if self._app is None:
+            self._app = Application.new_app(domain_name, app_name, application_version=version)
+            self._app.name = self.name
+            self._app.add_module(self.new_subject_module())
+            for event in self.iter_events():
+                module = event.get_module()
+                for study_form in event.iter_forms():
+                    study_form.add_form_to_module(module)
+                self._app.add_module(module)
+        return self._app
```
Comment (millerdev) [inline custom/openclinica/management/commands/odm_to_app.py]: This `self._app` member appears to be unused. It's not referenced outside of `get_app()`, and `get_app()` is only called once in the `Command` at the bottom of the file.

File: custom/openclinica/management/commands/odm_to_app.py [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+            yield StudyForm(form_def, self.meta)
+
+    def get_module(self):
+        """
+        Return a CommCare module
+        """
+        if self._module is None:
+            self._module = Module.new_module(self.name, None)
+            self._module.case_type = 'event'
+        return self._module
```
Comment (millerdev) [inline custom/openclinica/management/commands/odm_to_app.py]: `self._module` appears similar to `<Study>._app`. Is it needed?

File: custom/openclinica/management/commands/odm_to_app.py [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+
+    def add_form_to_module(self, module):
+        """
+        Add a CommCare form based in this form's item groups and items to a given CommCare module
+        """
+        if self._form is None:
+            self._form = module.new_form(self.name, None)
+            self._form.source = self.build_xform()
+        else:
+            module.add_insert_form(None, self._form)
```
Comment (snopoke) [inline custom/openclinica/management/commands/odm_to_app.py]: `add_form_to_module` only seems to be called once per instance so I don't think it needs the `if else`

**Comment by millerdev** [comment]
Reviewed. Looking good.

**Comment by kaapstorm** [comment]
Fixed case management and choice questions.

Here's the first form from the ODM document for the Captopril Study:
odm_app_form_builder

I will remove the "Open for review" label once #8953 is merged; it's waiting for some tests.

File: corehq/form_processor/models.py [hunk 5]
```diff
@@ -36,7 +36,24 @@ def is_saved(self):
...
+    def get_attachment_meta(self, attachment_name):
+        if hasattr(self, 'unsaved_attachments'):
+            for attachment in self.unsaved_attachments:
+                if attachment.name == attachment_name:
+                    return attachment
+        elif self.is_saved():
+            return self.attachments.filter(name=attachment_name).first()
+
+
+class XFormInstanceSQL(PreSaveHashableMixin, models.Model, RedisLockableMixIn, AttachmentMixin, AbstractXFormInstance):
```
Comment (dimagimon) [inline corehq/form_processor/models.py]: E501 line too long (119 > 115 characters)

File: corehq/form_processor/models.py [hunk 6]
```diff
@@ -289,7 +308,7 @@ class XFormPhoneMetadata(jsonobject.JsonObject):
     location = GeoPointProperty()
 
 
-class CommCareCaseSQL(PreSaveHashableMixin, models.Model, AbstractCommCareCase, RedisLockableMixIn, SaveStateMixin):
+class CommCareCaseSQL(PreSaveHashableMixin, models.Model, RedisLockableMixIn, AttachmentMixin, AbstractCommCareCase):
```
Comment (dimagimon) [inline corehq/form_processor/models.py]: E501 line too long (117 > 115 characters)

File: corehq/messaging/smsbackends/apposit/views.py [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+
+class AppositIncomingView(IncomingBackendView):
+    urlname = 'apposit_incoming'
+
+    def get(self, request, api_key, *args, **kwargs):
+        return HttpResponseBadRequest("ERROR: Expected POST")
+
+    def post(self, request, api_key, *args, **kwargs):
+        fromAddress = request.POST.get('fromAddress')
+        toAddress = request.POST.get('toAddress')
```
Comment (dimagimon) [inline corehq/messaging/smsbackends/apposit/views.py]: F841 local variable 'toAddress' is assigned to but never used

**Comment by kaapstorm** [comment]
This PR is to merge the odm_to_app branch into the xform_builder branch, which is what it's build on top of. The xform_builder branch is now merged into master, so I'm going to close this PR, and open a new one off master. Sorry about the churn.

**Pull request closed without merging**